velox/functions/prestosql/tests/utils/FunctionBaseTest.h

```diff
@@ -155,53 +156,109 @@ class FunctionBaseTest : public testing::Test,
     return std::dynamic_pointer_cast<T>(results[0]);
   }
 
-  // Evaluate the given expression once, returning the result as a std::optional
-  // C++ value. Arguments should be referenced using c0, c1, .. cn.  Supports
-  // integers, floats, booleans, and strings.
-  //
-  // Example:
-  //   std::optional<double> exp(std::optional<double> a) {
-  //     return evaluateOnce<double>("exp(c0)", a);
-  //   }
-  //   EXPECT_EQ(1, exp(0));
-  //   EXPECT_EQ(std::nullopt, exp(std::nullopt));
-  template <typename ReturnType, typename... Args>
-  std::optional<ReturnType> evaluateOnce(
+  /// Evaluate a given expression over a single row of input returning the
+  /// result as a std::optional C++ value. Prefer to use the `evaluateOnce()`
+  /// helper methods below when testing simple functions instead of manually
+  /// handling input and output vectors.
+  ///
+  /// Arguments should be referenced using c0, c1, .. cn.  Supports integers,
+  /// floats, booleans, strings, timestamps, and other primitive types.
+  ///
+  /// The template parameter type `TReturn` is mandatory and used to cast the
+  /// returned value. Always use the C++ type for the returned value (e.g.
+  /// `double` not `DOUBLE()`. If the function returns a custom type, use the
+  /// physical type that represent the custom type (e.g. `CustomType<T>::type`).
+  /// For example, use `int64_t` to return TIMESTAMP_WITH_TIME_ZONE()
+  ///
+  /// Input parameters and output should always wrapped in an std::optional to
+  /// allow the representation of null values.
+  ///
+  /// Example:
+  ///
+  ///   std::optional<double> exp(std::optional<double> a) {
+  ///     return evaluateOnce<double>("exp(c0)", a);
+  ///   }
+  ///   EXPECT_EQ(1, exp(0));
+  ///   EXPECT_EQ(std::nullopt, exp(std::nullopt));
```

Comment:
When testing default-null-behavior simple functions, it is redundant to test null inputs because these are handled by the framework, not by the function. It would be nice to clarify here that null tests are needed only when testing functions with non-default null behavior. It would also be nice to provide a version of evaluateOne that doesn't require std::optional.

Reply:
I guess here the point is less to test that the default null behavior works (which must already be covered elsewhere), but that the function null behavior is correctly defined. 

For example, if at some point someone changes a function definition and that changes (breaks) its semantic regarding nulls, you should have a test case to point that out. So IMHO, even if the function has default null behavior, you should still add a null in, null out check.